Tests/Algorithm/AlgorithmAddDataTests.cs

```diff
@@ -171,6 +174,412 @@ public void OnEndOfTimeStepDoesNotThrowWhenSeedsSameUnderlyingForTwoSecurities()
             Assert.AreEqual(data.Price, 2);
         }
 
+        //[TestCase("EURUSD", typeof(PsychSignalSentimentData), SecurityType.Cfd, false, false)]
+        //[TestCase("BTCUSD", typeof(PsychSignalSentimentData), SecurityType.Crypto, false, false)]
+        //[TestCase("CL", typeof(PsychSignalSentimentData), SecurityType.Future, false, false)]
+        //[TestCase("EURUSD", typeof(PsychSignalSentimentData), SecurityType.Forex, false, false)]
+        [TestCase("AAPL", typeof(PsychSignalSentimentData), SecurityType.Equity, true, true)]
+        public void AddDataSecuritySymbolWithUnderlying(string ticker, Type customDataType, SecurityType securityType, bool securityShouldBeMapped, bool customDataShouldBeMapped)
+        {
+            SymbolCache.Clear();
+            var qcAlgorithm = new QCAlgorithm();
+            qcAlgorithm.SubscriptionManager.SetDataManager(new DataManagerStub(qcAlgorithm));
+
+            Security asset;
+
+            switch (securityType)
+            {
+                case SecurityType.Cfd:
+                    asset = qcAlgorithm.AddCfd(ticker, Resolution.Daily);
+                    break;
+                case SecurityType.Crypto:
+                    asset = qcAlgorithm.AddCrypto(ticker, Resolution.Daily);
+                    break;
+                case SecurityType.Equity:
+                    asset = qcAlgorithm.AddEquity(ticker, Resolution.Daily);
+                    break;
+                case SecurityType.Forex:
+                    asset = qcAlgorithm.AddForex(ticker, Resolution.Daily);
+                    break;
+                case SecurityType.Future:
+                    asset = qcAlgorithm.AddFuture(ticker, Resolution.Daily);
+                    break;
+                default:
+                    throw new Exception($"SecurityType {securityType} is not valid for this test");
+            }
+
+            // Dummy here is meant to try to corrupt the SymbolCache. Ideally, SymbolCache should return non-custom data types with higher priority
+            // in case we want to add two custom data types, but still have them associated with the equity from the cache if we're using it.
+            // This covers the case where two idential data subscriptions are created.
+            var dummy = qcAlgorithm.AddData(customDataType, asset.Symbol, Resolution.Daily, qcAlgorithm.SubscriptionManager.Subscriptions.Where(x => x.SecurityType == securityType).Single().DataTimeZone);
```

Comment:
`Single` can accept a predicate, allowing you to replace the `Where` with `Single`